plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/Conversion.hs

```diff
@@ -0,0 +1,170 @@
+  AnyFormat DecimalFormat -> [T.pack $ toDecimal val]
+  _                       -> []
+
+alternateFracFormat :: Rational -> FormatType -> [Text]
+alternateFracFormat val fmt = case fmt of
+  AnyFormat DecimalFormat   -> [T.pack $ toFloatDecimal (fromRational val)]
+  FracFormat ExponentFormat -> [T.pack $ toFloatExpDecimal (fromRational val)]
+  FracFormat HexFloatFormat -> [T.pack $ toHexFloat (fromRational val)]
+  _                         -> []
+
+removeCurrentFormat :: Literal -> [FormatType] -> [FormatType]
+removeCurrentFormat lit fmts = case getSrcText lit of
+    Just src -> foldl (flip delete) fmts (sourceToFormatType src)
+    Nothing  -> fmts
+
+hexRegex, hexFloatRegex, binaryRegex, octalRegex, decimalRegex, numDecimalRegex :: Text
+hexRegex = "0[xX][a-fA-F0-9]+"
+hexFloatRegex = "0[xX][a-fA-F0-9]+\\.[a-fA-F0-9]+(p[+-]?[0-9]+)?"
+binaryRegex = "0[bB][0|1]+"
+octalRegex = "0[oO][0-8]+"
+decimalRegex = "\\d+"
+numDecimalRegex = "[0-9]+\\.[0-9]+[eE][+-]?[0-9]+"
+
+-- we export the regex for tests, but we want to be explicit in our matches
+-- so we need to match the beginning/end of the source text
+wrap :: Text -> Text
+wrap regex = "^" <> regex <> "$"
+
+sourceToFormatType :: Text -> [FormatType]
+sourceToFormatType srcText
+    | srcText =~ wrap hexRegex = [IntFormat HexFormat]
+    | srcText =~ wrap hexFloatRegex = [FracFormat HexFloatFormat]
+    | srcText =~ wrap octalRegex = [IntFormat OctalFormat]
+    | srcText =~ wrap binaryRegex = [IntFormat BinaryFormat]
+    -- can either be a NumDecimal or just a regular Fractional with an exponent
+    -- otherwise we wouldn't need to return a list
+    | srcText =~ wrap numDecimalRegex  = [IntFormat NumDecimalFormat, FracFormat ExponentFormat]
+    -- just assume we are in base 10 with no decimals
+    | otherwise = [AnyFormat DecimalFormat]
+
+-- grab the Numeric related extensions that are turned on for a file
+toFormatTypes :: [Extension] -> [FormatType]
+toFormatTypes =  (<>) [IntFormat HexFormat, IntFormat OctalFormat, FracFormat ExponentFormat, AnyFormat DecimalFormat]
+                 . mapMaybe (`lookup` numericPairs)
+
+-- current list of Numeric related extensions
+-- LexicalNegation --- 9.0.1 > --- superset of NegativeLiterals
+numericPairs :: [(Extension, FormatType)]
+numericPairs = [(NumericUnderscores, NoFormat), (NegativeLiterals, NoFormat)] <> intPairs <> fracPairs
+
+numericFormats :: [FormatType]
+numericFormats = map snd numericPairs
+
+intPairs :: [(Extension, FormatType)]
+intPairs = [(BinaryLiterals, IntFormat BinaryFormat), (NumDecimals, IntFormat NumDecimalFormat)]
+
+intFormats :: [FormatType]
+intFormats = map snd intPairs
+
+fracPairs :: [(Extension, FormatType)]
+fracPairs = [(HexFloatLiterals, FracFormat HexFloatFormat)]
+
+fracFormats :: [FormatType]
+fracFormats = map snd fracPairs
+
+-- Generate up to 3 possible choices where:
+-- dropWhile (\d -> val `div` d) > 1000) implies we want at MOST 3 digits to left of decimal
```

Comment:
```suggestion
-- | Generate up to 3 possible choices where:
-- dropWhile (\d -> val `div` d) > 1000) implies we want at MOST 3 digits to left of decimal
```

Reply:
This way it is haddock compliant and on hover you actually see the comment.

Additionally, check for valid Haddock syntax so it renders properly.